### ISSUE TITLE
ORA-01408: such column list already indexed  'CREATE INDEX subcategories_subcatid ON "MERCH"."INV_MJR"

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1000]
_Imported from trac ticket 1000.
Created by **rneville**
Opened in trac: Aug 16, 2011, 1:43:42 PM
Last modified in trac: Aug 31, 2011, 4:01:18 PM_
***
_by rneville:_
This is caused when the primary key index is also added to metadata as an Index().

Unique constraints cause the same problem.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 16, 2011, 3:25:30 PM:_

Ryan,
   For each table that has a primary key, Oracle has created an index for that primary key.  Those indexes should not have been included in changeset [[1646]](../commit/0acbdf802f13b544c30cd117c3713ecfcfeccc34), so you'll need to remove the redundant primary key index you added for each table.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 16, 2011, 3:45:46 PM:_

Ticket #997 has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 16, 2011, 3:46:07 PM:_

Ticket #998 has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 16, 2011, 3:46:26 PM:_

Ticket #999 has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 16, 2011, 3:47:00 PM:_

Ticket [#1001](1001) has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**rneville** trac comment on Aug 17, 2011, 12:03:41 PM:_

(Changeset [[1656]](../commit/cba112bc0d322e38d2b99d790f03ea3c85586819))
 * fixes #1000 index issues

**Comment by kbower** [comment]
_**rneville** trac status change on Aug 17, 2011, 12:03:41 PM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 17, 2011, 12:24:50 PM:_

(Changeset [[1658]](../commit/b7ee36942d2a9d39cbcdd7b50d9acc85bd069edc))
 * remove Index that was accidentally added with [[1656]](../commit/cba112bc0d322e38d2b99d790f03ea3c85586819), see #1000

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 17, 2011, 12:41:17 PM:_

Unique constraints cause the same problem.

**Comment by kbower** [comment]
_**kbower** trac status change on Aug 17, 2011, 12:41:17 PM:_ **reopened**

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 17, 2011, 1:18:47 PM:_

(Changeset [[1659]](../commit/fb6eb957595ffaecb999c7b81535214542743551))
 * change Unique constraints from Index, see #1000

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 17, 2011, 2:58:08 PM:_

[[1661]](../commit/3f45b1138ecb2469adfc0bbf7a36980e725a2465)

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 17, 2011, 3:06:10 PM:_

```
select '(' || all_ind_columns.table_name || ') ' || all_ind_columns.INDEX_NAME || ' ' || all_ind_columns.column_name from all_indexes, all_ind_columns where UNIQUENESS = 'UNIQUE' and 
all_indexes.table_name in  ('TRW_CD', 'COMM_CD', 'AR_TRN_ARC', 'CUST_ASP', 'INV_MNR', 'TRK_STOP', 'TET', 'SO_CMNT', 'INST_TERMS', 'AR_TRN_TP', 'ITM$SPIFF_CLNDR', 'AR_TRN', 'DISC_MNR', 'CUST2CUST_WARN', 'CUST_WARN', 'TAX_CD$TAT', 'PROD_GRP', 'PO_LN$ACTN_HST', 'TRK', 'TAX_CD', 'SO_LN', 'SO_ASP', 'EMP2STORE_GRP', 'STORE_GRP', 'EMP', 'EMP_SLSP', 'DEL', 'STORE', 'IST_LN', 'DISC', 'INST_INS_RT', 'DROP_CD', 'STORE_PHONE', 'PO_LN', 'IST_LN$SO_LN', 'ITM2RSKU', 'PACKAGE', 'ITM_TP', 'LOC_TP', 'ITM', 'ITM_CMNT', 'SPIFF_CD', 'ZIP', 'STORE_GRP$STORE', 'BILL_CYC', 'TAX_CHRT', 'INST_INS_RT_TERM', 'PO', 'ZIP2TAX', 'CUST_EMPL', 'VE', 'STORE$PRI', 'STORE$PROM_CLNDR', 'ASP', 'PUP', 'SO_LN$PO_LN', 'ITM_TRN', 'ZONE', 'FIN_PLAN_TP', 'INS_TP', 'ASP_PROMO', 'CUST', 'INV_MNR_CAT', 'ITM$RET_CLNDR', 'INV_MJR', 'ITM_FIFO', 'CNTR_SRT', 'ZIP2ZONE', 'TAT', 'LOC', 'ST', 'MOP', 'ITM$PROM_PRC_TP', 'VE_LEAD$MNR', 'ASSET_TP', 'INST_FIN_RT_CD', 'TRK_GRP2TRK', 'SO', 'ITM_FIFL', 'CUST_PL_EMPL', 'TRK_GRP', 'CUST_ASSET', 'CUST_PL_PREV_ADDR', 'INST_FIN_RT', 'CUST_SRT', 'IST', 'PROM_PRC_TP', 'INST_CNTR', 'SO_LN2DISC', 'CSH_DWR')
and not exists (select 0 from all_constraints where all_constraints.INDEX_NAME = all_indexes.INDEX_NAME and CONSTRAINT_TYPE = 'P')
and all_ind_columns.INDEX_NAME = all_indexes.INDEX_NAME
order by all_ind_columns.table_name, all_ind_columns.INDEX_NAME, all_ind_columns.column_position
```

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 17, 2011, 5:03:23 PM:_

Ticket #996 has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 17, 2011, 6:00:01 PM:_

(Changeset [[1663]](../commit/27b56284f7060f172d8286395725768ada1e2da1))
 * Add Mechanism to add indexes for non gers installations only (fixes [#1007](1007))
 * Remove another primary key index (fixes #1000)

**Comment by kbower** [comment]
_**kbower** trac status change on Aug 17, 2011, 6:00:01 PM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac status change on Aug 31, 2011, 3:24:43 PM:_ **reopened**

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 31, 2011, 3:26:56 PM:_

In at least one case, what was really two indexes were combined into one accidentally, causing problems (transfers table).

Also, UniqueConstraint()'s first argument must be a column, not name (string), or it is not added to the table's metadata.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 31, 2011, 3:54:24 PM:_

Interestingly, this led to the discover that Oracle may reuse an existing index when you `ALTER TABLE ADD CONSTRAINT ... UNIQUE(...)`

But sqlalchemy never reflects this constraint.

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 31, 2011, 4:01:18 PM:_

(Changeset [[1687]](../commit/66be96785fa16cb2b24e6ec5b4deb22c8f1837aa))
 * UniqueConstraint wasn't being used properly (see [#1050](1050) & #1000)
 * two "transfers" indexes where combined into one (fixes #1000)

**Comment by kbower** [comment]
_**kbower** trac status change on Aug 31, 2011, 4:01:18 PM:_ **closed**
